port glu to structured kernel

makslevental opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61433**

Differential Revision: [D29626288](https://our.internmc.facebook.com/intern/diff/D29626288)

Commit:
port glue to structured kernel

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626288).

makslevental commented on aten/src/ATen/native/GatedLinearUnit.cpp:
```diff
@@ -3,38 +3,44 @@
+DEFINE_DISPATCH(glu_backward_stub);
+
+TORCH_IMPL_FUNC(glu_out) (const Tensor& self, int64_t dim, const Tensor& out) {
+  auto selfSize = out.size(dim);
+  auto wrap_dim = maybe_wrap_dim(dim, self.dim());
+
   // half tensor
   Tensor firstHalf = self.narrow(wrap_dim, 0, selfSize);
   Tensor secondHalf = self.narrow(wrap_dim, selfSize, selfSize);
```
@ezyang is there a better way to do this? i.e. not recompute

makslevental commented on aten/src/ATen/native/GatedLinearUnit.cpp:
```diff
@@ -3,38 +3,44 @@
   auto newSizes = self.sizes().vec();
   newSizes[wrap_dim] = selfSize;
-  result.resize_(newSizes);
+  set_output(0, newSizes, {}, self.options(), {});
```
I'm not sure if we should have `build_unary_op(maybe_get_output(), self);` here after setting output. I had it here before but since we don't have `structured_inherits: TensorIteratorBase` in `native_functions.yaml` I had to remove it. Note that `TensorIteratorBase` can't be used (I think?) since `DECLARE_DISPATCH(activation_fn, glu_stub);` points to `using activation_fn = void (*)(TensorIterator&);`

makslevental replied:
also not sure if this is the right call to `set_output`; i see the 3 arg call in some places but I couldn't get type coercion/overloading to work for some reason when writing 

`set_output(0, newSizes, self.options())`

i.e. `newSizes` wasn't being coerced (cast?) to `IntArrayRef`(again that's my hypothesis for why 3 arg call didn't type check).

makslevental commented:
I tested this by running [TEST_F(FunctionalTest, GLU)](https://github.com/pytorch/pytorch/blob/master/test/cpp/api/functional.cpp#L981).

makslevental commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8322,14 +8322,15 @@
 
 - func: glu(Tensor self, int dim=-1) -> Tensor
+  structured_delegate: glu.out
+  device_check: NoCheck   # TensorIterator
```
not sure if this is correct. reading the docs on structured native functions (in particular `device_check`) it seems like `TensorIterator::borrowing_binary_op` should be performing this check already.

desertfire commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8322,14 +8322,15 @@
     CompositeExplicitAutograd: elu_
 
 - func: glu.out(Tensor self, int dim=-1, *, Tensor(a!) out) -> Tensor(a!)
+  structured: True
```
Do we need "structured_inherits: TensorIteratorBase" here?

makslevental replied:
i'm not sure but currently it's not possible due to

[`DECLARE_DISPATCH(activation_fn, glu_stub);`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Activation.h#L54) and [`using activation_fn = void (*)(TensorIterator&);
`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Activation.h#L16) in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Activation.h

makslevental committed:
Update on "port glu to structured kernel"


Differential Revision: [D29626288](https://our.internmc.facebook.com/intern/diff/D29626288)

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626288).

makslevental committed:
Update on "port glu to structured kernel"


Differential Revision: [D29626288](https://our.internmc.facebook.com/intern/diff/D29626288)

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626288).

makslevental committed:
Update on "port glu to structured kernel"


Differential Revision: [D29626288](https://our.internmc.facebook.com/intern/diff/D29626288)

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626288).

makslevental committed:
Update on "port glu to structured kernel"


Differential Revision: [D29626288](https://our.internmc.facebook.com/intern/diff/D29626288)

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626288).

desertfire commented on aten/src/ATen/native/GatedLinearUnit.cpp:
```diff
@@ -3,38 +3,38 @@
   // size output to half of input
   const int64_t selfSize = nIn / 2;
   auto newSizes = self.sizes().vec();
   newSizes[wrap_dim] = selfSize;
```
It seems like `newSizes` is not used at all?

makslevental replied:
ah right that was from before (when I was using `set_output`). I'll remove but I would like someone to confirm for me that indeed the substitution `set_output -> build_borrowing_binary_op` is correct.

ezyang replied:
Check if tests pass, but I think this is OK because the narrowed firstHalf/secondHalf tensors will induce TensorIterator to produce the correct size.

desertfire reviewed (approved):
LGTM. @ezyang, any comments?

desertfire commented on aten/src/ATen/native/GatedLinearUnit.cpp:
```diff
@@ -3,38 +3,38 @@
 
-Tensor& glu_out(const Tensor& self, int64_t dim, Tensor &result) {
+namespace meta {
+TORCH_META_FUNC(glu) (
```
Any reason you decided to change `glu_out` to `glu`? I think here the implementation is meant for `glu_out` and `glu` is meant to leverage this method.

makslevental replied:
That is the pattern here: https://github.com/pytorch/pytorch/pull/57619/files and indeed `TORCH_META_FUNC(glu_out)` does not correspond to a header declaration in `NativeMetaFunctions.h`. I am not sure if this is correct or if https://github.com/pytorch/pytorch/pull/57619/files is in error as well. @ezyang can probably advise.

ezyang replied:
Defining the meta func for glu and not glu_out is correct!

ezyang commented:
lgtm

makslevental committed:
Update on "port glu to structured kernel"


Differential Revision: [D29626288](https://our.internmc.facebook.com/intern/diff/D29626288)

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626288).

Krovatkin reviewed (approved):
Looks gut! Unfortunately, I'm not super familiar with structured kernels, so maybe you might want to wait for one more review.

makslevental committed:
Update on "port glu to structured kernel"


Differential Revision: [D29626288](https://our.internmc.facebook.com/intern/diff/D29626288)

[ghstack-poisoned]

makslevental committed:
Update on "port glu to structured kernel"


Differential Revision: [D29626288](https://our.internmc.facebook.com/intern/diff/D29626288)

[ghstack-poisoned]

makslevental committed:
Update on "port glu to structured kernel"


Differential Revision: [D29626288](https://our.internmc.facebook.com/intern/diff/D29626288)

[ghstack-poisoned]

makslevental committed:
Update on "port glu to structured kernel"


Differential Revision: [D29626288](https://our.internmc.facebook.com/intern/diff/D29626288)

[ghstack-poisoned]

makslevental committed:
Update on "port glu to structured kernel"


Differential Revision: [D29626288](https://our.internmc.facebook.com/intern/diff/D29626288)

[ghstack-poisoned]
